### PR TITLE
192 ensure slugs are unique for How-tos

### LINKED ISSUE #192: ensure slugs are unique
Have multiple 'test' docs which now can't be displayed as newer db entry with slug 'test' exists
child of #103

### PR DESCRIPTION
**Pull request by martisj**
Fixes #192  by displaying a message to the user after typing in the tutorial that the title must be unique. Since we generate a slug from the how-to title.

### CODE REVIEW

**Comment by martisj** [comment]
I had a quick look at trying to debounce network requests on validation, with no luck unfortunately. Final form is a bit of a beast to wrestle with for me. Anybody have input on how to solve this?
At the moment onchange of the tutorial title their is a network request. So for every letter change... 🤔 not ideal.

**Comment by kopecms** [comment]
> I had a quick look at trying to debounce network requests on validation, with no luck unfortunately. Final form is a bit of a beast to wrestle with for me. Anybody have input on how to solve this?
> At the moment onchange of the tutorial title their is a network request. So for every letter change...  not ideal.

**Comment by kopecms** [comment]
sorry... I should start reading what's on buttons ;p

**Comment by kopecms** [comment]
set validation on onfocusout but I am not react guy

**Comment by martisj** [comment]
> set validation on onblur but I am not react guy

I was trying to get the onblur working. Couldn't find a good way with react-final-form. Do you know of a better way @kopecms ?

**Comment by kopecms** [comment]
you can set validateOnBlur on Form element (then validation will run on bulring any input) or https://codesandbox.io/s/kxxw4l0p9o looks like something that may work

File: src/pages/Howto/Content/CreateHowto/CreateHowto.tsx [hunk 1]
```diff
@@ -47,6 +49,14 @@ const isNumber = (value: any) => {
   return
 }
 
+const uniqueSlug = async (slug: string) => {
```
Comment (kopecms) [inline src/pages/Howto/Content/CreateHowto/CreateHowto.tsx]: database access should be done store

**Comment by chrismclarke** [comment]
I see debouncing field validation is an issue with react-final-form:
https://github.com/final-form/react-final-form/issues/369, I think onblur is probably the best way forward too, although if complicated an interim solution could just be to run the validation check manually on submit.

**Comment by martisj** [comment]
Thanks for the feedback Chris!
Let me have a try tomorrow. I haven't looked into the debouncing yet though

**Comment by martisj** [comment]
I resolved the conflict and merged `master`. And I need some help on the onBlur validation. Seems not-so-trivial to set up onblur validation and that would change for the whole form?

**Comment by martisj** [comment]
Turns out I simply needed to add the validateOnBlur flag to the form. @chrismclarke can you have a look at the changes now?

File: src/pages/Howto/Content/CreateHowto/CreateHowto.tsx (hunk 1 above)
Comment (martisj) [reply src/pages/Howto/Content/CreateHowto/CreateHowto.tsx]: I moved this function into Howto.store

**Comment by chrismclarke** [comment]
So I just tried cloned this pr branch, created 2 posts with the same slug and it allowed me to so. I'm guessing whatever functionality isn't working correctly. 

Just to confirm, did you do your own testing first and were things working as anticipated?

**Comment by martisj** [comment]
@chrismclarke that's unfortunate indeed 🤔, what title did you try with? My impression of the issue was that slugs are derived from the title and thus I added validation to the `title` field. Making sure the title is unique is kind of the side effect. 
### non unique title (i.e. slug)
invalid image

### unique title
screencapture-localhost-3000-how-to-create-2019-03-28-18_52_51

And finally a movie showing that I cannot submit the form when the slug already exists

Mar-28-2019 19-05-08

**Comment by chrismclarke** [comment]
Perfect, thanks makes my job a lot easier. You can see the post I tried as it's on the dev site but perhaps the git pull didn't work properly. In any case let's merge this in and can add to the automated testing procedure once live (manually testing on the how-to's is a real pain point at the moment, so again your screenshots appreciated!)

**Pull request merged**